Fix #16994 - Fixing the issue when assigning database to user with '_'

iifawzi opened the pull request:
Signed-off-by: Fawzi E. Abdulfattah <iifawzie@gmail.com>

### Description
The problem was that if multiple databases were chosen, they were not unescaped and will be added as they are to `mysql`.`db`, that's why they appear with the backslash in the privileges view

Fixes #16994

iifawzi committed:
Fixing the issue when assigning database to user with '_'

Signed-off-by: Fawzi E. Abdulfattah <iifawzie@gmail.com>

williamdes reviewed (approved):
Awesome !
I will test it and merge it with your second PR

williamdes commented:
Works fine, but the UI on the page after submit still escapes twice the char. I did not find a quick fix for that

williamdes commented on libraries/classes/Controllers/Server/PrivilegesController.php:
```diff
@@ -229,7 +229,7 @@ public function index(): void
                         ($hostname ?? ''),
                         ($tablename ?? ($routinename ?? '')),
-                        ($db_name ?? ''),
+                        (Util::unescapeMysqlWildcards($db_name) ?? ''),
```
This would probably fail at some point because of a null value
The fix I pushed as 7c17889d780fe839bfe12c66b5a8fe01f68bf503 is:
```suggestion
                        (Util::unescapeMysqlWildcards($db_name ?? '')),
```

iifawzi commented:
> Works fine, but the UI on the page after submit still escapes twice the char. I did not find a quick fix for that

Ahh, I didn't notice it. I've found a fix for this, by unescaping the name of the databases in this else: 

https://github.com/phpmyadmin/phpmyadmin/blob/8500a6c5a15117b68355ba966aa976cb1a31b951/libraries/classes/Server/Privileges.php#L3265-L3273

To: 

```PHP
        if (! is_array($dbname) && strlen($dbname) > 0) {
            $_params['dbname'] = $dbname;
            if (strlen($tablename) > 0) {
                $_params['tablename'] = $tablename;
            }
        } else {
            foreach ($dbname as &$db_name) {
                $db_name = Util::unescapeMysqlWildcards($db_name);
            }
            unset($db_name);
            $_params['dbname'] = $dbname;
        }
```

It wouldn't affect anything else, we can use a separate array maybe called `unescaped_dbname` and pass it to twig, instead of mutating the `dbname` array if it's not preferred. 
I've tested and both solutions worked. What do you think? @williamdes

williamdes commented:
> I've tested and both solutions worked. 

Thank you !

> It wouldn't affect anything else, we can use a separate array maybe called unescaped_dbname and pass it to twig, instead of mutating the dbname array if it's not preferred.

I think this would be correct and avoid any possible side effects, using the unescaped_dbname option

iifawzi commented:
> > I've tested and both solutions worked.
> 
> Thank you !
> 
> > It wouldn't affect anything else, we can use a separate array maybe called unescaped_dbname and pass it to twig, instead of mutating the dbname array if it's not preferred.
> 
> I think this would be correct and avoid any possible side effects, using the unescaped_dbname option

Alright, should I create another PR for this, or I can add the commit to this PR?

williamdes commented:
> > > I've tested and both solutions worked.
> > 
> > 
> > Thank you !
> > > It wouldn't affect anything else, we can use a separate array maybe called unescaped_dbname and pass it to twig, instead of mutating the dbname array if it's not preferred.
> > 
> > 
> > I think this would be correct and avoid any possible side effects, using the unescaped_dbname option
> 
> Alright, should I create another PR for this, or I can add the commit to this PR?

You can add one more commit and I will craft a merge for it even if there is no PR ;)

iifawzi commented:
@williamdes, I've added the commit https://github.com/iifawzi/phpmyadmin/commit/b116bd2116eb26d23bd307213072c3b443f673bc, it works correctly, but there were two approaches to handle the case if `$dbname` is not an array, I've decided to make a check whether it's an array or not, if it's an array use `$unescaped_dbname` otherwise use `$dbname` 
```PHP
'database' => (is_array($dbname) && count($dbname) > 1) ? $unescaped_dbname : $dbname,
```

The other choice is to assign `$dbname` to `$unescaped_dbname` from the first place in the if statement,  but I think it might be misleading, thus I went with the first approach. 

Also, the statement I've added in the else can be avoided by converting the main `if` statement to ` else if `, feel free to commit the changes if you wish. 

Thank you for following up on this.

iifawzi commented:
I've figured a better approach that could solve both, the main issue and the UI issue, by unescaping the array elements as in the case of single db selected, from the beginning before distributing this array to other functions, and this could be done by escaping the array from ` getDataForDBInfo() ` that receives the selected elements from the first place.  

I've force pushed the changes https://github.com/iifawzi/phpmyadmin/commit/4056f24655aea22489e760c7013dfbcb9c5ecfe8 , and both issues solved 

<img width="1140" alt="Screen Shot 2021-07-14 at 2 51 22 AM" src="https://user-images.githubusercontent.com/46695441/125543861-7cd3e165-6ef7-4ef5-a109-6020083b3772.png">
<img width="1108" alt="Screen Shot 2021-07-14 at 2 50 31 AM" src="https://user-images.githubusercontent.com/46695441/125543862-65825ddf-e087-4467-a3e5-2e578e4d9269.png">

Sorry for this mess, I just felt like we're doing something wrong be making multiple changes in multiple places, investigated more and found this fix :")

williamdes commented:
Thanks for your fix, but I can not confirm it passes my tests. Maybe you should open a PR because if I understand well the previous fix also needs to be reverted with this solution

### Before
![image](https://user-images.githubusercontent.com/7784660/125614812-3a0fe134-32a2-47a0-b2d1-6384bc7af9f9.png)

### After

![image](https://user-images.githubusercontent.com/7784660/125614824-d1a10796-bce3-49cd-88ea-4fe29f9caf31.png)


After there is no escaping at all and I think it is wrong
Maybe go for the twig solution ?

iifawzi commented:
Yes, if it's an accurate fix, the old one should be reverted, but I'm confused now, what do you mean by `there's no escaping at all` I think the backslash shown in the query is the wrong one - in the before image - , because this means that the grant is made on the table with the name that contains an escaping characters, which's wrong I think and will be added to `mysql`.`db` with the escaping characters. could you check the `mysql`.`db` table to check whether the db name is added correctly without the backslashes or not?  if it's shown without the backslashes so I'm mistaken. 

Anyway, the last change I've made is only unescaping the array elements, as in the case when single db name is selected (shown in the else statement), if you could test with single db selected, the same behavior is happening, the query will be shown unescaped. 

https://github.com/iifawzi/phpmyadmin/blob/4056f24655aea22489e760c7013dfbcb9c5ecfe8/libraries/classes/Server/Privileges.php#L2920-L2933

I'm asking because as shown in the code above, we were unescaping in the case if single db is selected, I've just unescaped also in the case of multiple db is selected, so logically it shouldn't make any difference. 

Here's the `mysql`.`db` table, it shows that the grant are added successfully on the correct db names with the current implementation. 
 
<img width="839" alt="Screen Shot 2021-07-14 at 1 51 52 PM" src="https://user-images.githubusercontent.com/46695441/125617528-1beb048f-0fbd-42a7-99c1-fc1b02b68dc3.png">

before, will lead to this: 
<img width="732" alt="Screen Shot 2021-07-14 at 1 56 14 PM" src="https://user-images.githubusercontent.com/46695441/125618048-e8da1410-8443-4cd7-9a37-aa1037f6733a.png">
You will notice that the grant is added to incorrect db name, it's added to the escaped name, which's wrong as per my understand. 

I might be missing something though

williamdes commented:
> there's no escaping at all

I mean the last implementation you sent is not good according to 
> The _ and % wildcards are permitted when specifying database names in GRANT statements that grant privileges at the database level (GRANT ... ON db_name.*). This means, for example, that to use a _ character as part of a database name, specify it using the \ escape character as \_ in the GRANT statement, to prevent the user from being able to access additional databases matching the wildcard pattern (for example, GRANT ... ON `foo\_bar`.* TO ...).

Source: https://dev.mysql.com/doc/refman/8.0/en/grant.html

> Yes, if it's an accurate fix, the old one should be reverted

The actual fix you proposed is good (the one actually on QA branch). It only escapes `_` using `\` once.


Very nice avatar by the way :)

iifawzi commented:
> Source: https://dev.mysql.com/doc/refman/8.0/en/grant.html

Ohh, ok It make sense now, but doesn't this mean that there's a bug in the case when single db is selected? 
![Screen Shot 2021-07-15 at 12 25 39 AM](https://user-images.githubusercontent.com/46695441/125700974-f62cf803-5798-4e1c-be2d-db1856c2e94a.png)
it's not escaped at all too 👀 . 

Also, do you still prefer adding a commit to this repo? with the twig fix? or creating a separate PR?

> Very nice avatar by the way :)

Thank you! taken on the last day of my fourth college year, one year is remaining though 🤦🏻

williamdes commented:
> Also, do you still prefer adding a commit to this repo? with the twig fix? or creating a separate PR?

This is taking much more space than expected, to have a better overview a new PR would be better :)

> Ohh, ok It make sense now, but doesn't this mean that there's a bug in the case when single db is selected?

Yes, I think this is a bug, can you create an user to confirm the bug ?

iifawzi commented:
> Yes, I think this is a bug, can you create an user to confirm the bug ?

Yes, the same 
![image](https://user-images.githubusercontent.com/46695441/125704380-57c288c0-a2be-4f22-b4d0-ae62bde81211.png)

I think, this's the cause of the error, I didn't tested yet. 
https://github.com/phpmyadmin/phpmyadmin/blob/a0410aa77d903bec9ed72ec26e317094cd66d6e1/libraries/classes/Server/Privileges.php#L2929

> This is taking much more space than expected, to have a better overview a new PR would be better :)

Alright, I will create a separate PR with the twig fix discussed above, and the fix of the single db issue, if solved.

iifawzi commented:
Changes are almost done, but I've noticed something that I couldn't understand.
is `|escape_mysql_wildcards` intended?
https://github.com/phpmyadmin/phpmyadmin/blob/2f59c9b3bee42577edd09d659df03b7c0db3cc79/templates/server/privileges/privileges_summary.twig#L59

Why do we escape the already escaped names? if the database name is `test_db`, the code above will lead to `test_\\db` because it escaped twice, by the backend and by twig. 

if it's not intended and made by mistake, All the issues can be solved will less code, and without these changes everywhere. 
Thank you for your patience @williamdes, it's almost done, I've prepared the code for both ways :")

williamdes commented:
> Changes are almost done, but I've noticed something that I couldn't understand.
is |escape_mysql_wildcards intended?

Yes, I was searching for it, it is a twig extension we added (https://github.com/phpmyadmin/phpmyadmin/blob/QA_5_1/libraries/classes/Twig/UtilExtension.php#L165)
That could be super helpful meaning you would have less code to change

> if it's not intended and made by mistake, All the issues can be solved with less code and in a more cleaner way.
Thank you for your patience @williamdes, it's almost done, I've prepared the code for both ways :)

Let's go ;)

iifawzi commented:
I didn't get it :(, is it accurate to escape the db name twice? it's already escaped from the backend `escaped_databases[loop.index0]`, why escaping it again using the twig extension?
https://github.com/phpmyadmin/phpmyadmin/blob/2f59c9b3bee42577edd09d659df03b7c0db3cc79/templates/server/privileges/privileges_summary.twig#L59

williamdes commented:
> I didn't get it :(, is it accurate to escape the db name twice? it's already escaped from the backend `escaped_databases[loop.index0]`, why escaping it again using the twig extension?
> https://github.com/phpmyadmin/phpmyadmin/blob/2f59c9b3bee42577edd09d659df03b7c0db3cc79/templates/server/privileges/privileges_summary.twig#L59

Oh snap, I was thinking it was to "un" escape the wildcards
If you need to **un**escape you may add one more twig filter

> is it accurate to escape the db name twice?

No, it makes no sense. But maybe there is one case where it is not escaped 🤔 

Hey, look what I found when searching about this line: #16477
Maybe the mess comes from #16481 ..

iifawzi commented:
> No, it makes no sense. But maybe there is one case where it is not escaped
Hmm, if this's the case, it means that it will lead double escaping sometime, Anyway, I will test that my changes are working as expected for all the tests that I could think about, and then I will create a PR

> Maybe the mess comes from #16481.  
it seems like it made some conflicts, I will try to investigate more

williamdes commented:
> it seems like it made some conflicts, I will try to investigate more

Feel free to completely revert the fix and the make a fix for the fixed issue and the GUI + issue already fixed today by your first fix

Thank you so much for your work!
